### ISSUE TITLE
Add edge ID attribute to nx.readwrite.json_graph.cytoscape_data(G)

### ISSUE THREAD

**Issue by Wuelle** [networkx/networkx#4516]
Cytoscape.js requires edges to have an `id` attribute. (At least thats what i think, cy.json({elements: ...}) doesnt work when the new edges do not have an `id` attribute)
No such attribute is generated when callling `nx.readwrite.json_graph.cytoscape_data(G)` (The edges only have `source`, `target` and `weight` attributes)
```python
>>> import networkx as nx
>>> G = nx.Graph()
>>> G.add_weighted_edges_from([("a", "b", 1), ("b", "c", 2)])
>>> nx.readwrite.json_graph.cytoscape_data(G)
{'data': [], 'directed': False, 'multigraph': False, 'elements': {'nodes': [{'data': {'id': 'a', 'value': 'a', 'name': 'a'}}, {'data': {'id': 'b', 'value': 'b', 'name': 'b'}}, {'data': {'id': 'c', 'value': 'c', 'name': 'c'}}], 'edges': [{'data': {'weight': 1, 'source': 'a', 'target': 'b'}}, {'data': {'weight': 2, 'source': 'b', 'target': 'c'}}]}}
```

**Comment by dschult** [comment]
The [cytoscape tutorial demo data information](https://js.cytoscape.org/#notation/elements-json) says that an id is automatically assigned if not provided for some element. What do you mean by "doesnt work when new edges do not have an `id` attribute". It is possible that they have changed their requirements, but it'd be nice to look up the requirements before trying to adjust our code. The tutorial was easy for me to find, but perhaps there is a better authority. :)

**Comment by Wuelle** [comment]
Hm. This seems to only be an issue with the `cy.json` function. The potential Bug is demonstrated ![here](https://jsbin.com/xefukoc/1/edit?html,js,console,output). I say potential because  `cy.json` also updates the data of existing edges which are identified by their id, meaning that it doesnt really make sense for an edge not to have an id. I am not really sure whether this should be considered a cytoscape.js or a networkx issue.

**Comment by dschult** [comment]
If there documentation says that an id is automatically created when one is not supplied and it doesn't then it is a cytoscape issue. But the place I saw that was in a comment in the code of their tutorial demo.  Perhaps looking at their documentation more closely -- i.e. beyond the tutorial -- would tell us how it is supposed to work.

**Comment by Wuelle** [comment]
According to @maxkfranz this is intentional. Its basically like i assumed earlier, in `cy.json()` the Graph elements are updated declaratively and are identified by their ID, therefore it does not make sense for edges not to have an ID.

**Comment by dschult** [comment]
Yes, it is clear that Cytoscape assigns an ID for each edge. That's not my question. The tutorial says that each edge is given an ID if one is not provided. NetworkX doesn't provide one. So Cytoscape should give them an ID... right?

Maybe just start over from the top -- what is the trouble you are having?

**Comment by Wuelle** [comment]
Okay. 
I am trying to update my cytoscape graph with data exported from networkx. This is done using `cy.json()`. While the tutorial states that the ID attribute may be omitted, this **does not** apply to `cy.json()` (See [this issue](https://github.com/cytoscape/cytoscape.js/issues/2805)). Therefore, i cannot simply export my data from networkx using `nx.readwrite.json_graph.cytoscape_data(G)` , i instead have to do something like this:
```py
cytoscape_graph = nx.readwrite.json_graph.cytoscape_data(G)

# Assign a random id to every edge
for edge in cytoscape_graph["elements"]["edges"]:
    edge["data"]["id"] = secrets.token_urlsafe(32)

```
I think something like this should be done in `nx.readwrite.json_graph.cytoscape_data(G)`.
After all, the whole point of that function is to avoid having to process the data before passing it to cytoscape.js.

**Comment by maxkfranz** [comment]
There are two cases for using `cy.json()`:

(1) One-and-done import:  You can use `cy.json()` to import a network from an external source, say an HTTP fetch or from NetworkX.  In this case, you're done at the point of the `cy.json()` call:  Everything is imported.  You don't care that you use autogenerated edge IDs on the Cytoscape side.

(2) Declaratively updating a network:  You can use `cy.json(jsonSpec)` to update the network so that it is exactly as specified in `jsonSpec`.  Cytoscape will do diffs in order to update the network.  Elements that are new are added, and elements that are no longer in the latest `jsonSpec` are removed.  This can't work unless element IDs are provided in order to perform the diff.

So it seems that the approach that NetworkX is currently using works just fine for (1), but it doesn't work out of the box with (2).  If NetworkX were to provide edge IDs for `cytoscape_data`, then both (1) and (2) would work seamlessly.

Does that help to clarify things?

**Comment by dschult** [comment]
Thanks for these descriptions. I think I understand better now: The issue is trying to **update** a cytoscape graph from data in NetworkX. Unless the edge ids are stored on the NetworkX graph there is no way to know what edge ids are being used on the cytoscape side and that means you can't update the cytoscape graph from networkx.

One solution/workaround is to do as @wuelle demonstrates: add your own edge ids before moving the graph to cytoscape.  Then the edge ids are part of the networkx graph and can be used for future updates.

The proposed better solution is to add code to the `cytoscape_data` function. This code would check if every edge has an id. If so, it proceeds as usual. Otherwise it creates an edge id for each edge that doesn't have one and stores it as a networkx edge attribute. Then it proceeds as usual.

I'm guessing that means we need a new keyword argument to provide the name of the attribute that represents the id. We also need to figure out a good way to assign the id values. Are integers reasonable choices? I'm assuming they need to be unique within the network. Any danger if they are not unique across two graphs? We have stayed away from edge ids due to problems like this (and storage costs for very large graphs). We refer to edges using the nodes rather than an id.  Can Cytoscape handle edge ids that are just a tuple of the nodes?

**Comment by maxkfranz** [comment]
>  Are integers reasonable choices? I'm assuming they need to be unique within the network. Any danger if they are not unique across two graphs? We have stayed away from edge ids due to problems like this (and storage costs for very large graphs). We refer to edges using the nodes rather than an id. Can Cytoscape handle edge ids that are just a tuple of the nodes?

IDs are strings, while ints are accepted and casted to strings.  You can use a string tuple, e.g. `(foo, bar)`, to identify simple edges.  Node ID tuples are insufficient for multigraphs.

**Comment by dschult** [comment]
OK... If I understand correctly then, we would be fine using the following code to set an "id" attribute for each edge.  Indeed, perhaps we should simply include this as an example in the doc_string so people:
1) know how to do it
2) don't have to do it
3) can adapt it to what they really need.

```python
nx.set_edge_attributes(G, {e: str(e) for e in G.edges()}, "id")
```

Then the next question (never ending questions).  Isn't this also true for node ID?  If you want to update a node property via `cy.json` don't you need to have the node ids also?

What else does CytoScape need?

**Comment by maxkfranz** [comment]
For this use case, you’re right:  All elements should have unique IDs.

**Comment by dschult** [comment]
Do the node IDs have to be unique within the nodes, or over all element type IDs? 
Are there any IDs other than those for nodes and edges?

**Comment by maxkfranz** [comment]
> Do the node IDs have to be unique within the nodes, or over all element type IDs?

Over all element type IDs

> Are there any IDs other than those for nodes and edges?

No

**Comment by dschult** [comment]
I guess that means we can't use node strings for the IDs and node-tuple strings for the edge IDs because the graph with nodes {0, 1, (0, 1)} and edge (0, 1) will have node (0, 1) and also edge (0, 1).

Why would anyone do that?  But still... there it is... an example of why we don't supply edge IDs.
:)               Let me think about how to choose the edge IDs so they are more informative than an integer count of the elements...

Still... for almost all cases, the following will work:
```python
nx.set_node_attributes(G, {n: str(n) for n in G.nodes}, "id")
nx.set_edge_attributes(G, {e: str(e) for e in G.edges}, "id")
```
(untested)

**Comment by rossbar** [comment]
Having reviewed the above, my feeling is that this should be handled by the user rather than baked into the `cytoscape_json` function, i.e. something similar to [@Wuelle 's](https://github.com/networkx/networkx/issues/4516#issuecomment-758535520) or [@dschult 's](https://github.com/networkx/networkx/issues/4516#issuecomment-760533871) depending on the specific case.
